Lab1: Complete the TODO parts in the following code. 
- Using California Housing Dataset from sklearn, select input attributes 1,3,4  as the input features. 
- Using K-fold cross validation technique (https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.KFold.html), complete the implementation to train a regression model and report performance merics when asked in the following code. 
- For multiple degrees of model complexity (i.e., degree of polynomial in this exercise) in a for-loop, obtain the model with the minimum reducible_error, polynomial degree, and run the obtained model on the test data. For this part,you should use the split the data into train and test by [75:25] rate and report mse of the final model on test data. 
- Analyse the results of model performance according to different degrees of polynomial and the number of folds used. You can manipulate the code and share your analysis in terms of the performance of the model (mse and total error), such as for instnace which degree of the model complexity (in relation to the polynomial order) would give a better model? Feel free to include other analysis about the generated models in relation to their performance results. You can event plot the results to support your analysis. 

In [ ]:
import numpy as np
import pandas as pd
import seaborn as sns
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import PolynomialFeatures
from sklearn.datasets import fetch_california_housing
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as pltfrom sklearn.model_selection import KFold


In [ ]:
def polynomial_regression(degree, X, y, folds, test_size=0.25, random_state=None):
    # Define number of folds for cross-validation
    kf = KFold(folds)

    # Initialize lists to store results for variance, bias2s, total_error, and models
    variances = []
    bias_sq = []
    total_errors = []
    models = []

    # Set the polynomial degree of the model
    poly_features = PolynomialFeatures(degree)
    X_poly = poly_features.fit_transform(X)

    # Perform cross-validation
    for train_index, test_index in kf.split(X_poly):
        # Split data into training and testing sets for this fold
        X_train, X_test = X_poly[train_index], X_poly[test_index]
        y_train, y_test = y[train_index], y[test_index]

        # Fit polynomial regression model
        model = LinearRegression()
        model.fit(X_train, y_train)

        # Make predictions on the test set
        y_pred = model.predict(X_test)

        # Calculate variance and bias for this fold
        variance = np.var(y_pred)
        mean_predictions = np.mean(y_pred)
        bias2 = np.mean((mean_predictions - y_test) ** 2)

        total_error = bias2 + variance

        
        # Append results to lists
        variances.append(variance)
        bias_sq.append(bias2)
        total_errors.append(total_error)
        models.append(model)

        # Print results for this fold
        print("Variance: {:.4f}, Bias2: {:.4f}, Total error: {:.4f}".format(variance, bias2, bias2 + variance))

    # print the total_error of the best model
    min_error_index = np.argmin(total_errors)
    best_model = models[min_error_index]
    best_total_error = total_errors[min_error_index]
    print(f"Best Model Total Error: {best_total_error:.4f}")


    

    # Testing the final model on the test data
    X_train, X_test, y_train, y_test = train_test_split(X_poly, y, test_size=test_size, random_state=42)

    # Obtain the predictions on the test data
    y_pred_best = best_model.predict(X_test)
    
    # store mse score of the model applied on the test data
    mse = mean_squared_error(y_test, y_pred_best)
    

    return print(f"MSE:{mse} Best Model:{best_model}")



#### Data Schema
    - MedInc        median income in block group
    - HouseAge      median house age in block group
    - AveRooms      average number of rooms per household
    - AveBedrms     average number of bedrooms per household
    - Population    block group population
    - AveOccup      average number of household members
    - Latitude      block group latitude
    - Longitude     block group longitude

    WANT TO PREDICT 
    - MedHouseVal   median house value

data: each district in California

In [51]:

# Example usage: load California Housing Dataset and select the first, third, and forth attributes as input features in X
cali_houses = fetch_california_housing(as_frame=True)
X = cali_houses['data']
y = cali_houses['target']

# selecting the first,third, and fourth attributes
X_selected = pd.DataFrame(X , columns = ["MedInc", "AveRooms", "AveBedrms"])
df_y = pd.DataFrame(y , columns = cali_houses.target_names)
df = pd.concat([X_selected , df_y] , axis=1)

# show concatenated data, X_selected, and Y 
display(df)
display(X_selected)
df_y

,MedInc,AveRooms,AveBedrms,MedHouseVal
0,8.3252,6.984127,1.023810,4.526
1,8.3014,6.238137,0.971880,3.585
2,7.2574,8.288136,1.073446,3.521
3,5.6431,5.817352,1.073059,3.413
4,3.8462,6.281853,1.081081,3.422
...,...,...,...,...
20635,1.5603,5.045455,1.133333,0.781
20636,2.5568,6.114035,1.315789,0.771
20637,1.7000,5.205543,1.120092,0.923
20638,1.8672,5.329513,1.171920,0.847


,MedInc,AveRooms,AveBedrms
0,8.3252,6.984127,1.023810
1,8.3014,6.238137,0.971880
2,7.2574,8.288136,1.073446
3,5.6431,5.817352,1.073059
4,3.8462,6.281853,1.081081
...,...,...,...
20635,1.5603,5.045455,1.133333
20636,2.5568,6.114035,1.315789
20637,1.7000,5.205543,1.120092
20638,1.8672,5.329513,1.171920


,MedHouseVal
0,4.526
1,3.585
2,3.521
3,3.413
4,3.422
...,...
20635,0.781
20636,0.771
20637,0.923
20638,0.847


ERROR:tornado.general:Uncaught exception in ZMQStream callback
Traceback (most recent call last):
  File "/opt/anaconda3/envs/FM/lib/python3.10/site-packages/zmq/eventloop/zmqstream.py", line 565, in _log_error
    f.result()
  File "/opt/anaconda3/envs/FM/lib/python3.10/site-packages/ipykernel/kernelbase.py", line 341, in dispatch_control
    await self.process_control(msg)
  File "/opt/anaconda3/envs/FM/lib/python3.10/site-packages/ipykernel/kernelbase.py", line 347, in process_control
    idents, msg = self.session.feed_identities(msg, copy=False)
  File "/opt/anaconda3/envs/FM/lib/python3.10/site-packages/jupyter_client/session.py", line 994, in feed_identities
    raise ValueError(msg)
ValueError: DELIM not in msg_list
ERROR:tornado.general:Uncaught exception in ZMQStream callback
Traceback (most recent call last):
  File "/opt/anaconda3/envs/FM/lib/python3.10/site-packages/zmq/eventloop/zmqstream.py", line 565, in _log_error
    f.result()
  File "/opt/anaconda3/envs/FM/lib/pytho

In [50]:
degrees = range(1, 6)  # Try polynomial degrees from 1 to 5
# Try degrees from 1 to 5 and in a loop, report mse of the best model trained using k-fold cross validation and print("Degree:", degree, "MSE:", mse)
for degree in degrees: 
    print(f"\nTrying model at degree of {degree}")
    polynomial_regression(degree, X_selected, y, 4, test_size=0.25, random_state=None)



Trying model at degree of 1
Variance: 0.6158, Bias2: 1.1677, Total error: 1.7835
Variance: 0.7082, Bias2: 1.4063, Total error: 2.1145
Variance: 0.6127, Bias2: 1.0820, Total error: 1.6946
Variance: 0.7005, Bias2: 1.5872, Total error: 2.2876
Best Model Total Error: 1.6946
MSE:0.6625085543605442 Best Model:LinearRegression()

Trying model at degree of 2
Variance: 1.2405, Bias2: 1.1594, Total error: 2.3999
Variance: 0.6276, Bias2: 1.3972, Total error: 2.0248
Variance: 0.7275, Bias2: 1.0818, Total error: 1.8093
Variance: 0.7402, Bias2: 1.5897, Total error: 2.3299
Best Model Total Error: 1.8093
MSE:0.6313771231504414 Best Model:LinearRegression()

Trying model at degree of 3
Variance: 10.2368, Bias2: 1.1570, Total error: 11.3937
Variance: 0.6542, Bias2: 1.3948, Total error: 2.0490
Variance: 0.7933, Bias2: 1.0827, Total error: 1.8760
Variance: 0.7835, Bias2: 1.5909, Total error: 2.3744
Best Model Total Error: 1.8760
MSE:0.6178020253509674 Best Model:LinearRegression()

Trying model at degree

/opt/anaconda3/envs/FM/lib/python3.10/site-packages/sklearn/linear_model/_base.py:280: RuntimeWarning: divide by zero encountered in matmul
  return X @ coef_ + self.intercept_
/opt/anaconda3/envs/FM/lib/python3.10/site-packages/sklearn/linear_model/_base.py:280: RuntimeWarning: overflow encountered in matmul
  return X @ coef_ + self.intercept_
/opt/anaconda3/envs/FM/lib/python3.10/site-packages/sklearn/linear_model/_base.py:280: RuntimeWarning: invalid value encountered in matmul
  return X @ coef_ + self.intercept_
/opt/anaconda3/envs/FM/lib/python3.10/site-packages/sklearn/linear_model/_base.py:280: RuntimeWarning: divide by zero encountered in matmul
  return X @ coef_ + self.intercept_
/opt/anaconda3/envs/FM/lib/python3.10/site-packages/sklearn/linear_model/_base.py:280: RuntimeWarning: overflow encountered in matmul
  return X @ coef_ + self.intercept_
/opt/anaconda3/envs/FM/lib/python3.10/site-packages/sklearn/linear_model/_base.py:280: RuntimeWarning: invalid value encountered

### Commentary
After using 4 fold cross validation across in a for-loop to test polynomial regression degrees of one to five we see some interesting patterns. Firstly we see that the fifth degree polynomial regression produces the lowest MSE of 0.5917~ and see that the variance is quite unstable across the folds with one fold reaching a variance of 149979.5824. The fifth degree polynomial regression also produces the highest best model total error of 2.0663. The first degree polynomial resgression has the highest MSE of 0.6625~ which is comparitively not much higher than the lowest MSE and has the lowest model total error. Also looking at the first degree polynomial regression's  variance across folds we see that its a lot more stable when compared to the higher degree polynomial regression model. In regards to the safest and most stable model I would say that it would be the first-degree polynomial regression model.